hazelcast/src/hazelcast/client/serialization/pimpl/DataInput.cpp

```diff
@@ -143,22 +146,18 @@ namespace hazelcast {
                 }
 
                 std::auto_ptr<std::string> DataInput::readUTF() {
-                    int32_t len = readInt();
-                    if (util::Bits::NULL_ARRAY == len) {
-                        return std::auto_ptr<std::string>(NULL);
+                    int32_t charCount = readInt();
+                    if (util::Bits::NULL_ARRAY == charCount) {
+                        return std::auto_ptr<std::string>();
                     } else {
-                        int numBytesToRead = 0;
-                        for (int i = 0; i < len ; ++i) {
-                            checkAvailable(1);
-                            int numBytesForChar = getNumBytesForUtf8Char(&buffer[pos] + numBytesToRead);
-                            numBytesToRead += numBytesForChar;
-                            checkAvailable(numBytesToRead);
+                        utfBuffer.reserve((size_t) MAX_UTF_CHAR_SIZE * charCount);
+                        byte b;
+                        for (int i = 0; i < charCount; ++i) {
+                            b = readByte();
+                            util::UTFUtil::readUTF8Char(*this, b, utfBuffer);
```

Comment:
Is there a specific reason why we diverged from java implementation here? 
The `readUTF8Char` seems more simple there.
And why did we create `ByteReadable` interface? If it is just for testing I think we need to complicate tests not code itself if necessary.